Allow modes `z` and `Z` for docker volumes

bernt-matthias opened the pull request:
Fixes https://github.com/galaxyproject/galaxy/issues/18989

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

bernt-matthias committed:
allow z and Z for docker volumes

bernt-matthias committed:
allow any ordering in the mode string

bernt-matthias committed:
remove unnecessary comprehension

bernt-matthias committed:
also check for z in the re-duplication code

hexylena commented on lib/galaxy/tool_util/deps/container_volumes.py:
```diff
@@ -7,7 +7,7 @@
 
 class ContainerVolume(metaclass=ABCMeta):
-    valid_modes = frozenset({"ro", "rw"})
+    valid_modes = frozenset({"ro", "rw", "z"})
```
```suggestion
    valid_modes = frozenset({"ro", "rw", "z", "Z"})
```

bernt-matthias committed:
Update lib/galaxy/tool_util/deps/container_volumes.py

Co-authored-by: Helena <hexylena@galaxians.org>

bernt-matthias committed:
fix logic

bernt-matthias committed:
restructure to make roundtrip testable

The combination

- preprocess_volumes which creates Volume objects and creates strings
  again in the end and
- creation of DockerVolumes which are only temporary and are soon
  converted to str again

confused me and needed testing IMO. Also code was redundant.

hexylena reviewed (approved):
Thanks for fixing this @bernt-matthias
